### PR TITLE
[FSTORE-698] Write Hudi performance best practices guide

### PR DESCRIPTION
**Pull request by moritzmeister**

### CODE REVIEW

**Comment by SirOibaf** [comment]
Can you add:
```
---
description: blabla bla
---
```
at the beginning of the page to add the page description. See https://hopsworks.atlassian.net/browse/HWORKS-421

**Comment by moritzmeister** [comment]
@SirOibaf jap sorry, I knew I forgot something. paid attention to not wrap the image in a link though ;)

**Review by jimdowling — CHANGES_REQUESTED** [review]
Mostly typos. 
But, maybe you should add something about row group sizes. 
You could say, for example, that they should be aligned with the HopsFS block size (128MB, by default), as recommended here:

https://medium.com/swlh/insights-into-parquet-storage-ac7e46b94ffe#:~:text=Row%20Group%20%3A%20It%20is%20a,to%20the%20HDFS%20block%20size.&text=Page%3A%20Column%20chunks%20are%20divided%20up%20into%20pages.


Write something like. 
Unless you change your Hopsworks configuration to change the default HopsFS block size, then you won't need to worry about configuring the row group size.


Note: clients can read in parallel from different row groups in a parquet file, hence why it is important for configuration.

File: docs/user_guides/fs/feature_group/create.md [hunk 1]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+If the inserted Dataframe contains only a single feature group partition, this feature group partition will be written with five parquet files.
+If the inserted Dataframe contains multiple feature group partitions, the parquet files will be split among those partition, potentially more parquet files will be added.
+
+<figure markdown>
+  ![feature group partitioning](../../../assets/images/guides/feature_group/fg-partition-files.png)
+  <figcaption>Mapping in-memory partitions to tasks, workers, executors and feature group partition files for a feature group insert</figcaption>
+</figure>
+
+!!! tip "Setting shuffle parallelism"
+    In practice that means the shuffle parallelism should be set equal to the number of feature group partitions being contained in the inserted dataframe.
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md:220]: I am always unsure, in pandas it's Dataframe, in spark it's DataFrame

File: docs/user_guides/fs/feature_group/create.md [hunk 2]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+    This will create one parquet file per feature group partition, which in many cases is optimal.
+
+    Theoretically, this rule holds up to a partition size of 2GB, which is the limit of Spark. However, one should bump this up accordingly already for smaller inputs.
+    We recommend having shuffle parallelism `hoodie.[insert|upsert|bulkinsert].shuffle.parallelism` such that its at least input_data_size/500MB.
```
Comment (tdoehmen) [inline docs/user_guides/fs/feature_group/create.md:224]: I wonder how this interacts with the `write.parquet.max.file.size` property (https://hudi.apache.org/docs/configurations/#writeparquetmaxfilesize)

File: docs/user_guides/fs/feature_group/create.md [hunk 3]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+#### Insert Best Practices
+
+When designing a feature group, it is worth taking a look at how this feature group will be queried in the future, in order to optimize it for those query patterns.
+At the same time, Spark and Hudi tend to overpartition writes, slowing down the write.
+The best practices described in this section hold both for the Streaming API and the Batch API.
+
+There are three main considerations that influence the write and also the query performance:
+
+1. Partitioning on feature group level
+2. Parquet file size within a feature group partition
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md:138]: But that's not something you can control in hudi? so not really worth adding in this list I think?

File: docs/user_guides/fs/feature_group/create.md [hunk 4]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
 
 By using partitioning the system will write the feature data in different subdirectories, thus allowing you to write 10240 files per partition.
 
+#### Streaming Write API
+
+As explained above, the stream parameter controls whether to enable the streaming write APIs to the online and offline feature store.
+For Python environments, the stream pipeline is enabled by default, and this is the recommended way of writing to a feature group from
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md]: yeah, markdown doesn't really care about line breaks in the source file, so sometimes I just add a line break to make it more readable for us. But it doesn't affect the final html.

File: docs/user_guides/fs/feature_group/create.md [hunk 5]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+        description="Weather Features",
+        online_enabled=True,
+        primary_key=['location_id'],
+        partition_key=['day'],
+        event_time='event_time',
+        stream=True
+    )
+    ```
+
+When using the streaming API, the data will be written to the online storage (if `online_enabled=True`) on the shortest path. However, you can control when the sync to
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md]: changed it simply to "will be written directly to the online storage..."

File: docs/user_guides/fs/feature_group/create.md [hunk 6]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+
+# start the backfill for all three inserts
+# note the job object is always the same, you don't need to call it three times
+job.run()
+```
+
+#### Insert Best Practices
+
+When designing a feature group, it is worth taking a look at how this feature group will be queried in the future, in order to optimize it for those query patterns.
+At the same time, Spark and Hudi tend to overpartition writes, slowing down the write.
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md]: tried to make it more clear, let me know what you think

File: docs/user_guides/fs/feature_group/create.md [hunk 7]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+The best practices described in this section hold both for the Streaming API and the Batch API.
+
+There are three main considerations that influence the write and also the query performance:
+
+1. Partitioning on feature group level
+2. Parquet file size within a feature group partition
+3. Backfilling of feature group partitions
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md:137]: see comment above

File: docs/user_guides/fs/feature_group/create.md [hunk 8]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+    end_time=end_time,
+    description='Description of a dataset',
+)
+```
+
+Assuming, the feature group was partitioned by a daily event time column, for example the features are updated with a daily batch job, the feature store will only have to list and read the files in the directories of those
+six months that are being queried.
+
+!!! danger "Too granular event time columns"
+    An event time column which is too granular, such as a timestamp, shouldn't be used as partition key.
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md:170]: tried to clarify with an example

File: docs/user_guides/fs/feature_group/create.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md]: so when I am not referring to the concept of a dataframe in a specific framework I tend to write just "dataframe"

File: docs/user_guides/fs/feature_group/create.md [hunk 9]
```diff
@@ -101,6 +78,178 @@ MaxDirectoryItemsExceededException - The directory item limit is exceeded: limit
...
+    That means if you feature pipeline is written to write one feature group partition at a time,
+    you cannot run it multiple times in parallel for backfilling.
+
+    The recommended approach is to unionise the dataframes and insert them with a single `fg.insert()` instead.
+    For stream enabled feature groups, it is enough to defer starting the backfill job after multiple inserts,
+    [as described above](#streaming-write-api).
+
 ### Register the metadata and save the feature data
 
 The snippet above only created the metadata object on the Python interpreter running the code. To register the feature group metadata and to save the feature data with Hopsworks, you should invoke the `save` method:
```
[the remark this answers is not in the document]
Comment (moritzmeister) [reply docs/user_guides/fs/feature_group/create.md]: thanks, this was old docs, fixed it.

**Review by jimdowling — APPROVED** [review]

File: docs/user_guides/fs/feature_group/create.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimdowling) [reply docs/user_guides/fs/feature_group/create.md:136]: If you can't control it in Hudi, then leave it out.

**Pull request merged**